File: leetcode/1971. Find if Path Exists in Graph/main.py
from collections import defaultdict

from typing import List, Dict, Set


class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph: Dict[int, List[int]] = defaultdict(list)
        for edge in edges:
            u, v = edge[0], edge[1]
            graph[u].append(v)
            graph[v].append(u)

        return self.dfs(graph, source, destination, set())

    def dfs(self, graph: Dict[int, List[int]], current: int, target: int, visited: Set[int]) -> bool:
        if current == target:
            return True

        if current in visited:
            return False
        else:
            visited.add(current)

        for next in graph[current]:
            if self.dfs(graph, next, target, visited):
                return True

        return False

File: leetcode/1971. Find if Path Exists in Graph/main.rs
use std::collections::{HashMap, HashSet};

impl Solution {
    pub fn valid_path(n: i32, edges: Vec<Vec<i32>>, source: i32, destination: i32) -> bool {
        let mut graph: HashMap<i32, Vec<i32>> = HashMap::new();
        for edge in edges {
            let (u, v) = (edge[0], edge[1]);
            graph.entry(u).or_insert(vec![]).push(v);
            graph.entry(v).or_insert(vec![]).push(u);
        }

        Self::dfs(&graph, source, destination, &mut HashSet::new())
    }

    fn dfs(
        graph: &HashMap<i32, Vec<i32>>,
        current: i32,
        target: i32,
        visited: &mut HashSet<i32>,
    ) -> bool {
        if current == target {
            return true;
        }

        if visited.contains(&current) {
            return false;
        } else {
            visited.insert(current);
        }

        if let Some(neighbours) = graph.get(&current) {
            for &next in neighbours {
                if Self::dfs(graph, next, target, visited) {
                    return true;
                }
            }
        }

        false
    }
}
